cmd/restic/cmd_dump.go

```diff
@@ -74,56 +74,138 @@ func splitPath(p string) []string {
+	filteredSet := make(map[string][]string)
+
+	for _, p := range pathComponentsList {
+		splittedPath := splitPath(path.Clean(p))
+
+		isCovered := false
+		// Check if any prefix of the path exists in the pathSet
+		for i := len(splittedPath) - 1; i >= 1; i-- {
+			prefixPath := path.Join(splittedPath[0:i]...)
+			if _, exists := pathSet[prefixPath]; exists {
+				isCovered = true
+				break
+			}
+		}
+		if !isCovered {
+			pathStr := path.Join(splittedPath...)
+			filteredSet[pathStr] = splittedPath
+		}
+	}
+
+	// Convert the values of the set to a list
+	filteredList := make([][]string, 0, len(filteredSet))
+	for _, splittedPath := range filteredSet {
+		filteredList = append(filteredList, splittedPath)
+	}
+
+	return filteredList
+}
+
+func filterPathComponents(pathComponentsList [][]string, prefix string) [][]string {
+	var filteredList [][]string
+	for _, pathComponents := range pathComponentsList {
+		if len(pathComponents) > 1 && pathComponents[0] == prefix {
+			filteredList = append(filteredList, pathComponents[1:])
+		}
+	}
+	return filteredList
+}
+
+func printFromTree(ctx context.Context, tree *restic.Tree, repo restic.BlobLoader, prefix string, pathComponentsList [][]string, d *dump.Dumper, canWriteArchiveFunc func() error) error {
 	// If we print / we need to assume that there are multiple nodes at that
 	// level in the tree.
-	if pathComponents[0] == "" {
+	if pathComponentsList[0][0] == "" {
 		if err := canWriteArchiveFunc(); err != nil {
 			return err
 		}
 		return d.DumpTree(ctx, tree, "/")
 	}
 
-	item := filepath.Join(prefix, pathComponents[0])
-	l := len(pathComponents)
+	// make a set of the nodes that we dumped
+	dumpedNodes := make(map[string]struct{})
+
 	for _, node := range tree.Nodes {
 		if ctx.Err() != nil {
 			return ctx.Err()
 		}
 
-		// If dumping something in the highest level it will just take the
-		// first item it finds and dump that according to the switch case below.
-		if node.Name == pathComponents[0] {
-			switch {
-			case l == 1 && node.Type == restic.NodeTypeFile:
-				return d.WriteNode(ctx, node)
-			case l > 1 && node.Type == restic.NodeTypeDir:
-				subtree, err := restic.LoadTree(ctx, repo, *node.Subtree)
-				if err != nil {
-					return errors.Wrapf(err, "cannot load subtree for %q", item)
-				}
-				return printFromTree(ctx, subtree, repo, item, pathComponents[1:], d, canWriteArchiveFunc)
-			case node.Type == restic.NodeTypeDir:
-				if err := canWriteArchiveFunc(); err != nil {
-					return err
+	out:
+		// iterate over all pathComponents
+		for _, pathComponents := range pathComponentsList {
+			item := filepath.Join(prefix, pathComponents[0])
+			l := len(pathComponents)
+
+			// If dumping something in the highest level it will just take the
+			// first item it finds and dump that according to the switch case below.
+			if node.Name == pathComponents[0] {
+				switch {
+				case l == 1 && node.Type == restic.NodeTypeFile:
+					d.WriteNode(ctx, node)
```

Comment:
This behaves incorrectly when a user specifies multiple file paths. `WriteNode` only dumps the file content without any metadata. Also a combination of a directory + file path results in a corrupted output.